### PR TITLE
Sanitize filepath from URL

### PR DESCRIPTION
**Pull request by johnwparent**
Some urls (particularly from MS Downloads) contain characters than cannot be directly parsed into filepaths on Windows (ex. '?') and cause an error when the local filepath for a given archive is created. This PR sanitizes the url string before it's converted into a filepath.

### CODE REVIEW

File: lib/spack/spack/stage.py [hunk 1]
```diff
@@ -388,7 +387,7 @@ def save_filename(self):
             # This prefers using the URL associated with the default fetcher if
             # available, so that the fetched resource name matches the remote
             # name
-            return possible_filenames[0]
+            return sup.ensure_legal_path(possible_filenames[0])
```
Comment (scheibelp) [inline lib/spack/spack/stage.py]: Can this move to just formatting `os.path.basename(self.default_fetcher.url)` in `expected_archive_files`? I think that is more clear about where the problem is occurring (while what you did here should technically always work, it doesn't always need to be applied to `possible_filenames[0]`).

You could also apply it to other paths in that function but e.g. we generate `.mirror_paths` ourselves, so it isn't the same problem as using the URL basename.

File: lib/spack/spack/util/path.py [hunk 2]
```diff
@@ -87,6 +87,21 @@ def path_to_os_path(*pths):
     return ret_pths
 
 
+def ensure_legal_path(pth):
```
Comment (scheibelp) [inline lib/spack/spack/util/path.py]: Two requests:

* Can you rename this to `sanititize_file_path` (when I see a function with the prefix `ensure_` it is often a check that doesn't actually produce a result)
* Can you add a docstring mentioning that this formats strings that are intended to be used as paths that will be created by Spack? In particular I think it's worth clarifying we mean "legal" in the sense that it is a valid name for a file on the filesystem

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a couple minor requests about organization and naming. Let me know if they seem off-track. Thanks!

**Comment by scheibelp** [comment]
Also: where did you get the restrictions on path names? The best source I found was https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations - it might be worth adding a reference to that in a comment (or whatever source you are using if you found a better one).

**Comment by scheibelp** [comment]
Adding a unit test would also be helpful here.

**Comment by johnwparent** [comment]
> Also: where did you get the restrictions on path names? The best source I found was https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations - it might be worth adding a reference to that in a comment (or whatever source you are using if you found a better one).

This is also the resource I used. I will include it.

File: lib/spack/spack/stage.py (hunk 1 above)
Comment (johnwparent) [reply lib/spack/spack/stage.py]: Sure, the less we touch the smaller the potential for issues later, I'll make the change.

File: lib/spack/spack/util/path.py (hunk 2 above)
Comment (johnwparent) [reply lib/spack/spack/util/path.py]: Will do in both instances.

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks!

**Comment by haampie** [comment]
Sorry @scheibelp @johnwparent, but once more I'm not impressed with new Windows code paths affecting unix.

What's the point here.

On unix a path is a sequence of bytes, with `/` separators, and `\0` is not allowed (cause C-strings).

What I see here is: split on `/`, remove `/` from each component (?!), join with `/`. Please...

Also, it doesn't guard against null bytes if the idea was to do more than applying the identity function:

```python
In [5]: with open(spack.util.path.sanitize_file_path("\0"), "w") as f: 
   ...:     f.write('hello') 
   ...:                                                                                                                               
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/spack/lib/spack/spack/cmd/python.py in <module>
----> 1 with open(spack.util.path.sanitize_file_path("\0"), "w") as f:
      2     f.write('hello')
      3     

ValueError: embedded null byte
```

However, given that the input is trusted, there is simply no point of "sanitizing file paths" on unix. Please drop it.

File: lib/spack/spack/util/path.py [hunk 3]
```diff
@@ -87,6 +87,42 @@ def path_to_os_path(*pths):
...
+
+    Args:
+        pth: string containing path to be created
+            on the host filesystem
+
+    Return:
+        sanitized string that can legally be made into a path
+    """
+    # on unix, splitting path by seperators will remove
+    # instances of illegal characters on join
```
Comment (haampie) [inline lib/spack/spack/util/path.py:106]: ???

**Pull request merged**